eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/ReplaceableTxPublisher.scala

```diff
@@ -73,485 +75,270 @@ object ReplaceableTxPublisher {
-                                     watcher: ActorRef[ZmqWatcher.Command],
-                                     context: ActorContext[ReplaceableTxPublisher.Command],
-                                     timers: TimerScheduler[ReplaceableTxPublisher.Command],
-                                     loggingInfo: TxPublishLogContext)(implicit ec: ExecutionContext = ExecutionContext.Implicits.global) {
+private class ReplaceableTxPublisher(nodeParams: NodeParams, bitcoinClient: BitcoinCoreClient, watcher: ActorRef[ZmqWatcher.Command], context: ActorContext[ReplaceableTxPublisher.Command], timers: TimerScheduler[ReplaceableTxPublisher.Command], loggingInfo: TxPublishLogContext)(implicit ec: ExecutionContext = ExecutionContext.Implicits.global) {
 
   import ReplaceableTxPublisher._
-  import nodeParams.{channelKeyManager => keyManager}
 
   private val log = context.log
 
   def start(): Behavior[Command] = {
     Behaviors.receiveMessagePartial {
-      case Publish(replyTo, cmd, targetFeerate) =>
-        cmd.txInfo match {
-          case _: ClaimLocalAnchorOutputTx => checkAnchorPreconditions(replyTo, cmd, targetFeerate)
-          case htlcTx: HtlcTx => checkHtlcPreconditions(replyTo, cmd, htlcTx, targetFeerate)
-          case claimHtlcTx: ClaimHtlcTx => checkClaimHtlcPreconditions(replyTo, cmd, claimHtlcTx, targetFeerate)
-        }
+      case Publish(replyTo, cmd) => checkPreconditions(replyTo, cmd)
       case Stop => Behaviors.stopped
     }
   }
 
-  def checkAnchorPreconditions(replyTo: ActorRef[TxPublisher.PublishTxResult], cmd: TxPublisher.PublishReplaceableTx, targetFeerate: FeeratePerKw): Behavior[Command] = {
-    // We verify that:
-    //  - our commit is not confirmed (if it is, no need to claim our anchor)
-    //  - their commit is not confirmed (if it is, no need to claim our anchor either)
-    //  - our commit tx is in the mempool (otherwise we can't claim our anchor)
-    val commitTx = cmd.commitments.fullySignedLocalCommitTx(nodeParams.channelKeyManager).tx
-    val fundingOutpoint = cmd.commitments.commitInput.outPoint
-    context.pipeToSelf(bitcoinClient.isTransactionOutputSpendable(fundingOutpoint.txid, fundingOutpoint.index.toInt, includeMempool = false).flatMap {
-      case false => Future.failed(CommitTxAlreadyConfirmed)
-      case true =>
-        // We must ensure our local commit tx is in the mempool before publishing the anchor transaction.
-        // If it's already published, this call will be a no-op.
-        bitcoinClient.publishTransaction(commitTx)
-    }) {
-      case Success(_) => PreconditionsOk
-      case Failure(CommitTxAlreadyConfirmed) => CommitTxAlreadyConfirmed
-      case Failure(reason) if reason.getMessage.contains("rejecting replacement") => RemoteCommitTxPublished
-      case Failure(reason) => UnknownFailure(reason)
-    }
+  def checkPreconditions(replyTo: ActorRef[TxPublisher.PublishTxResult], cmd: TxPublisher.PublishReplaceableTx): Behavior[Command] = {
```

Comment:
I fail to see the benefit of creating additional level of actors. I understand the goal of splitting `ReplaceableTxPublisher` into several smaller parts, but why not separate the individual behavior methods themselves? Does it facilitate testing somehow?

Reply:
It's mostly about making it easier to read.
Otherwise we have a lot more logic in the same file, and a lot more `Command`s, which make it harder to reason about.
Especially when you're trying to figure out which commands should be handled in each state.
That's why I prefer to isolate between different actors instead of just different behaviors in the same actor, because each actor's commands are better separated.

Reply:
You could split the behaviors in several files though, and you could also separate the behaviors commands:

```scala

sealed trait Command
sealed trait Step1 extends Command
...
sealed trait Step2 extends Command
...

def step1: Behavior[Step1] = ???

def step2: Behavior[Step2] = ???
```

I would say that this allows to quickly understand the flow by looking at the `Command` definitions, and still physically separate implementations. But I don't have a strong opinion and haven't yet used that pattern myself.

Reply:
It wouldn't work well with the pattern I use where I delay some commands (to handle in a future state) with `timers.startSingleTimer`, those would need to use the `Command` trait...I think it's always a cleaner and stricter separation to use a separate actor, and here I was getting lost in the complexity without that refactoring...

Reply:
Fair enough!

Reply:
Somewhat related: WDYT about putting as much args as possible in the actor's `private class`? That's the purpose right? It makes individual behaviors much leaner and makes their requirements more obvious, see: 8039cba8ba29a29b06f6beaaeffd0fe0aa0b8845.

I understand that you may consider that `replyTo` is request-specific and shouldn't be in the actor's state, but I think akka's doc sees the private class as just a container to [prevent passing around too many parameters](https://doc.akka.io/docs/akka/current/typed/style-guide.html#passing-around-too-many-parameters).

Reply:
That makes sense, there are a few arguments that I'm passing around all the time, they make sense to be in the `private class` fields :+1:

Reply:
I believe that the other actors from the `publish` package would benefit from the same refactoring.

Reply:
Done in https://github.com/ACINQ/eclair/pull/2113/commits/fa3e17d3319663377da63f3e5f7a5b52fae38b37

Reply:
Nit:

```patch
diff --git a/eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/MempoolTxMonitor.scala b/eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/MempoolTxMonitor.scala
index 44146d65e..75f810eec 100644
--- a/eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/MempoolTxMonitor.scala
+++ b/eclair-core/src/main/scala/fr/acinq/eclair/channel/publish/MempoolTxMonitor.scala
@@ -86,7 +86,7 @@ private class MempoolTxMonitor(nodeParams: NodeParams, cmd: MempoolTxMonitor.Pub
         waitForConfirmation()
       case PublishFailed(reason) if reason.getMessage.contains("rejecting replacement") =>
         log.info("could not publish tx: a conflicting mempool transaction is already in the mempool")
-        sendResult(cmd.replyTo, TxRejected(cmd.tx.txid, TxRejectedReason.ConflictingTxUnconfirmed))
+        sendResult(TxRejected(cmd.tx.txid, TxRejectedReason.ConflictingTxUnconfirmed))
       case PublishFailed(reason) if reason.getMessage.contains("bad-txns-inputs-missingorspent") =>
         // This can only happen if one of our inputs is already spent by a confirmed transaction or doesn't exist (e.g.
         // unconfirmed wallet input that has been replaced).
@@ -94,21 +94,21 @@ private class MempoolTxMonitor(nodeParams: NodeParams, cmd: MempoolTxMonitor.Pub
         Behaviors.same
       case PublishFailed(reason) =>
         log.error("could not publish transaction", reason)
-        sendResult(cmd.replyTo, TxRejected(cmd.tx.txid, TxRejectedReason.UnknownTxFailure))
+        sendResult(TxRejected(cmd.tx.txid, TxRejectedReason.UnknownTxFailure))
       case status: InputStatus =>
         if (status.spentConfirmed) {
           log.info("could not publish tx: a conflicting transaction is already confirmed")
-          sendResult(cmd.replyTo, TxRejected(cmd.tx.txid, TxRejectedReason.ConflictingTxConfirmed))
+          sendResult(TxRejected(cmd.tx.txid, TxRejectedReason.ConflictingTxConfirmed))
         } else if (status.spentUnconfirmed) {
           log.info("could not publish tx: a conflicting mempool transaction is already in the mempool")
-          sendResult(cmd.replyTo, TxRejected(cmd.tx.txid, TxRejectedReason.ConflictingTxUnconfirmed))
+          sendResult( TxRejected(cmd.tx.txid, TxRejectedReason.ConflictingTxUnconfirmed))
         } else {
           log.info("could not publish tx: one of our wallet inputs is not available")
-          sendResult(cmd.replyTo, TxRejected(cmd.tx.txid, TxRejectedReason.WalletInputGone))
+          sendResult(TxRejected(cmd.tx.txid, TxRejectedReason.WalletInputGone))
         }
       case CheckInputFailed(reason) =>
         log.error("could not check input status", reason)
-        sendResult(cmd.replyTo, TxRejected(cmd.tx.txid, TxRejectedReason.TxSkipped(retryNextBlock = true))) // we act as if the input is potentially still spendable
+        sendResult(TxRejected(cmd.tx.txid, TxRejectedReason.TxSkipped(retryNextBlock = true))) // we act as if the input is potentially still spendable
       case Stop =>
         Behaviors.stopped
     }
@@ -132,7 +132,7 @@ private class MempoolTxMonitor(nodeParams: NodeParams, cmd: MempoolTxMonitor.Pub
         if (nodeParams.minDepthBlocks <= confirmations) {
           log.info("txid={} has reached min depth", cmd.tx.txid)
           context.system.eventStream ! EventStream.Publish(TransactionConfirmed(loggingInfo.channelId_opt.getOrElse(ByteVector32.Zeroes), loggingInfo.remoteNodeId, cmd.tx))
-          sendResult(cmd.replyTo, TxConfirmed(cmd.tx), Some(messageAdapter))
+          sendResult(TxConfirmed(cmd.tx), Some(messageAdapter))
         } else {
           Behaviors.same
         }
@@ -147,13 +147,13 @@ private class MempoolTxMonitor(nodeParams: NodeParams, cmd: MempoolTxMonitor.Pub
       case status: InputStatus =>
         if (status.spentConfirmed) {
           log.info("tx was evicted from the mempool: a conflicting transaction has been confirmed")
-          sendResult(cmd.replyTo, TxRejected(cmd.tx.txid, TxRejectedReason.ConflictingTxConfirmed))
+          sendResult(TxRejected(cmd.tx.txid, TxRejectedReason.ConflictingTxConfirmed))
         } else if (status.spentUnconfirmed) {
           log.info("tx was evicted from the mempool: a conflicting transaction replaced it")
-          sendResult(cmd.replyTo, TxRejected(cmd.tx.txid, TxRejectedReason.ConflictingTxUnconfirmed))
+          sendResult(TxRejected(cmd.tx.txid, TxRejectedReason.ConflictingTxUnconfirmed))
         } else {
           log.info("tx was evicted from the mempool: one of our wallet inputs disappeared")
-          sendResult(cmd.replyTo, TxRejected(cmd.tx.txid, TxRejectedReason.WalletInputGone))
+          sendResult(TxRejected(cmd.tx.txid, TxRejectedReason.WalletInputGone))
         }
       case CheckInputFailed(reason) =>
         log.error("could not check input status", reason)
@@ -164,9 +164,9 @@ private class MempoolTxMonitor(nodeParams: NodeParams, cmd: MempoolTxMonitor.Pub
     }
   }

-  def sendResult(replyTo: ActorRef[TxResult], result: TxResult, blockSubscriber_opt: Option[ActorRef[CurrentBlockCount]] = None): Behavior[Command] = {
+  def sendResult(result: TxResult, blockSubscriber_opt: Option[ActorRef[CurrentBlockCount]] = None): Behavior[Command] = {
     blockSubscriber_opt.foreach(actor => context.system.eventStream ! EventStream.Unsubscribe(actor))
-    replyTo ! result
+    cmd.replyTo ! result
     Behaviors.stopped
   }

```

Reply:
Good catch, I missed that one! Fixed in https://github.com/ACINQ/eclair/pull/2113/commits/441a8ee752d0ec285aca9db236e11cea1d69f199